crates/protocol/src/batch/core.rs

```diff
@@ -48,4 +51,82 @@ impl Batch {
             }
         }
     }
+
+    /// Attempts to encode the batch to a writer.
+    pub fn encode(&self, out: &mut Vec<u8>) -> Result<(), BatchEncodingError> {
```

Comment:
you could also use `out: &mut dyn bytes::BufMut` here then this accepts vec and bytesmut
but vec is fine